### PR TITLE
caclmgrd: monitor state_db to update dhcp acl

### PR DESCRIPTION
**Pull request by trzhang-msft**
caclmgrd: monitor mux_cable_table in state_db to update dhcp acl
- if the state changes to 'standby', add acl to block dhcp packets based on ingress interfaces
- if the state changes to 'active', delete acl
- if the state changes to 'unknown', also delete acl to avoid potential disconnect
- both addition and deletion follow checking the existence of the rules

The change has been verified on a virtual switch based testbed.

### CODE REVIEW

**Comment by lguohan** [comment]
missing unit test.

File: src/sonic-host-services/scripts/caclmgrd [hunk 1]
```diff
@@ -383,6 +463,10 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "ip6tables -A INPUT -p icmpv6 --icmpv6-type router-solicitation -j ACCEPT")
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "ip6tables -A INPUT -p icmpv6 --icmpv6-type router-advertisement -j ACCEPT")
 
+        # Add iptables commands to link the DCHP chain to block dhcp packets based on ingress interfaces
+        if self.DualToR:
+            iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -A INPUT -p udp --dport 67 -j DHCP")
```
Comment (prsunny) [inline src/sonic-host-services/scripts/caclmgrd:468]: Are the below rules required for dualtor? Also why not port 68?

**Review by prsunny — COMMENTED** [review]
Please provide details of the setting up dhcp chain/acl rule in the description. Need more clarifications on the approach taken.

File: src/sonic-host-services/scripts/caclmgrd [hunk 2]
```diff
@@ -337,11 +416,12 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -P FORWARD ACCEPT")
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -P OUTPUT ACCEPT")
 
-        # Add iptables command to flush the current rules
-        iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -F")
-
-        # Add iptables command to delete all non-default chains
-        iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -X")
+        # Add iptables command to flush the current rules and delete all non-default chains
```
[the remark this answers is not in the document]
Comment (tahmed-dev) [reply src/sonic-host-services/scripts/caclmgrd:419]: This is to shield DHCP related rule in separate chain list. When flushing, exclude DHCP chain and flush/delete all other chains other than built-in chains.

File: src/sonic-host-services/scripts/caclmgrd [hunk 3]
```diff
@@ -653,8 +752,21 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
...
+                if self.DualToR:
+                    while True:
+                        key, op, fvs = subscribe_mux_cable.pop()
+                        if not key:
+                            break
+                        self.log_info("mux cable update : '%s'" % str((key, op, fvs)))
+                        self.update_dhcp_acl(key, op, dict(fvs))
+                continue
+
+            ctrl_plane_acl_notification = set()
```
Comment (tahmed-dev) [inline src/sonic-host-services/scripts/caclmgrd:769]: what is the use of `ctrl_plane_acl_notification`?

File: src/sonic-host-services/scripts/caclmgrd [hunk 4]
```diff
@@ -319,6 +327,77 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
         else:
             return False
 
+    def setup_dhcp_chain(self, namespace):
+        all_chains = self.get_chain_list(self.iptables_cmd_ns_prefix[namespace], [""])
+        dhcp_chain_exist = 1 if "DHCP" in all_chains else 0
```
Comment (tahmed-dev) [inline src/sonic-host-services/scripts/caclmgrd:332]: what about using boolean?
`dhcp_chain_exist = "DHCP" in all_chains`

File: src/sonic-host-services/scripts/caclmgrd [hunk 5]
```diff
@@ -84,6 +86,8 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
 
     UPDATE_DELAY_SECS = 0.5
 
+    DualToR = 0
```
Comment (tahmed-dev) [inline src/sonic-host-services/scripts/caclmgrd:89]: boolean?

File: src/sonic-host-services/scripts/caclmgrd (hunk 1 above)
Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd:468]: this is to direct matching into DHCP chain, dhcp discovers/requests use dport 67

File: src/sonic-host-services/scripts/caclmgrd (hunk 3 above)
Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd:769]: this is the existing code moved down from L646

File: src/sonic-host-services/scripts/caclmgrd (hunk 4 above)
Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd]: updated

File: src/sonic-host-services/scripts/caclmgrd (hunk 5 above)
Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd]: updated

File: src/sonic-host-services/scripts/caclmgrd (hunk 2 above)
[the remark this answers is not in the document]
Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd:419]: yes, if caclmgrd restarts, we flush all.

**Comment by trzhang-msft** [comment]
retest this please

**Review by tahmed-dev — DISMISSED** [review]

File: src/sonic-host-services/scripts/caclmgrd [hunk 6]
```diff
@@ -653,8 +752,21 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
...
+
+            if db_id == state_db_id:
+                if self.DualToR:
+                    while True:
+                        key, op, fvs = subscribe_mux_cable.pop()
+                        if not key:
+                            break
+                        self.log_info("mux cable update : '%s'" % str((key, op, fvs)))
+                        self.update_dhcp_acl(key, op, dict(fvs))
+                continue
```
Comment (prsunny) [inline src/sonic-host-services/scripts/caclmgrd:767]: Can you confirm this flow is tested on a non-dualtor testbed?

Comment (trzhang-msft) [reply src/sonic-host-services/scripts/caclmgrd:767]: yes, i tested it with both subtype=dualtor or none on a vs testbed

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
Is unit test separate PR? or are you planning to add to this?

**Comment by prsunny** [comment]
> Is unit test separate PR? or are you planning to add to this?

Ok, I see this PR - https://github.com/Azure/sonic-buildimage/pull/8359. You could have it in same PR, FYI

**Review by tahmed-dev — APPROVED** [review]

**Pull request merged**